xxhash_avx512_amd64.s

```diff
@@ -0,0 +1,215 @@
+//go:build !appengine && gc && !purego && go1.22
+// +build !appengine
+// +build gc
+// +build !purego
+// +build go1.22
+
+#include "textflag.h"
+
+DATA ·initWideAvx512<>+0(SB)/8, $0x60ea27eeadc0b5d6
+DATA ·initWideAvx512<>+8(SB)/8, $0xc2b2ae3d27d4eb4f
+DATA ·initWideAvx512<>+16(SB)/8, $0x0000000000000000
+DATA ·initWideAvx512<>+24(SB)/8, $0x61c8864e7a143579
+GLOBL ·initWideAvx512<>(SB), NOSPLIT|NOPTR, $32
+
+#define p      SI
+#define h      AX
+#define end    DI
+#define temp   BX
+#define prime1 DX
+#define prime2 R8
+
+#define state   Y0
+#define xstate  X0
+#define yprime1 Y1
+#define yprime2 Y2
+#define ytemp   Y3
+#define xtemp   X3
+
+#define yround()                  \
+	VPMULLQ ytemp, yprime2, ytemp \
+	VPADDQ  ytemp, state, state   \
+	VPROLQ  $31, state, state     \
+	VPMULLQ state, yprime1, state \
+
+#define blockLoop(length)         \
+	MOVL    $0x1f, end            \
+	ANDNQ   length, end, end      \
+	ADDQ    p, end                \
+	PCALIGN $64                   \
+loop_32:                          \
+	VMOVDQU (p), ytemp            \
+	yround()                      \
+	ADDQ    $32, p                \
+	CMPQ    p, end                \
+	JNE     loop_32
+
+#define n      CX
+#define prime3 R9
+#define prime4 R10
+#define prime5 R11
+
+// lateMergeRound performs mergeRound on h given the value from round0
+#define lateMergeRound(v) \
+	XORQ  v, h            \
+	IMULQ prime1, h       \
+	ADDQ  prime4, h
+
+// func Sum64(b []byte) uint64
+// Requires: AVX, AVX2, AVX512DQ, AVX512F, AVX512VL, BMI
+TEXT ·Sum64(SB), NOSPLIT|NOFRAME, $0-32
+	CMPB ·useAvx512(SB), $0x00
+	JNE  do_avx512
+	JMP  ·sum64Scalar(SB)
+zero:
+	MOVQ $0xef46db3751d8e999, h
+	MOVQ h, ret+24(FP)
+	RET
+
+do_avx512:
+	MOVQ         b_base+0(FP), p
+	MOVQ         b_len+8(FP), n
+	MOVQ         $0x9e3779b185ebca87, prime1
+	MOVQ         $0xc2b2ae3d27d4eb4f, prime2
+	MOVQ         $0x165667b19e3779f9, prime3
+	MOVQ         $0x85ebca77c2b2ae63, prime4
+	MOVQ         $0x27d4eb2f165667c5, prime5
+
+	LEAQ         (prime5)(n*1), h // precompute h for the shortcuts
+	JCXZQ        zero
+	CMPQ         n, $3
+	JBE          loop_1
+	CMPQ         n, $7
+	JBE          do_4
+	CMPQ         n, $31
+	JBE          loop_8
+
+	VMOVDQU      ·initWideAvx512<>+0(SB), state
+	VPBROADCASTQ prime1, yprime1
```

Comment:
Turns out it doesn't look like it:

Instruction | Op1 | Op2 | Op3 | Op4 | masking operand | APMVol | Cpuid flag | Ops | Unit | Latency | Throughput
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
VPBROADCASTQ | ymm1 | reg64 |   |   | {k1}{z}/{k1} | 4 | AVX512 | 2 | FP1/2 | 1 | 1

I couldn't find a difference in benchmarks either.
uops.info doesn't agree and gives a variable cost but it says this is as fast an l1 memory load in the *worst case*. So I think it's better this way.
I think this is pipelined into `VMOVQ xstate, h` so that why I can't measure any difference either way.

I can confirm the YMM → GPR are expensive bellow through `VPEXTRQ` and `VMOVQ` but it's less than `YMM → mem (l1) → GPR`.